### PR TITLE
DNN-21973: retry the search when there have exception throw during search process.

### LINKED ISSUE #2183: sometimes search result will be empty in azure app service
sometimes search result will be empty in azure app service for some instances, such as if the app have 3 instances, one will not able to search but the other 2 instances works fine. there have error logged:

```
[ERROR] DotNetNuke.Services.Exceptions.Exceptions - System.IO.IOException: An unexpected network error occurred.
 
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.ReadCore(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.FileStream.Read(Byte[] array, Int32 offset, Int32 count)
   at Lucene.Net.Store.SimpleFSDirectory.SimpleFSIndexInput.ReadInternal(Byte[] b, Int32 offset, Int32 len)
   at Lucene.Net.Store.BufferedIndexInput.Refill()
   at Lucene.Net.Store.BufferedIndexInput.ReadByte()
   at Lucene.Net.Store.IndexInput.ReadVInt()
   at Lucene.Net.Index.TermBuffer.Read(IndexInput input, FieldInfos fieldInfos)
   at Lucene.Net.Index.SegmentTermEnum.Next()
   at Lucene.Net.Index.SegmentTermEnum.ScanTo(Term term)
   at Lucene.Net.Index.TermInfosReader.Get(Term term, Boolean useCache)
   at Lucene.Net.Index.TermInfosReader.Terms(Term term)
   at Lucene.Net.Index.DirectoryReader.MultiTermEnum..ctor(IndexReader topReader, IndexReader[] readers, Int32[] starts, Term t)
   at Lucene.Net.Index.DirectoryReader.Terms(Term term)
   at Lucene.Net.Search.PrefixTermEnum..ctor(IndexReader reader, Term prefix)
   at Lucene.Net.Search.PrefixQuery.GetEnum(IndexReader reader)
   at Lucene.Net.Search.MultiTermQuery.ConstantScoreAutoRewrite.Rewrite(IndexReader reader, MultiTermQuery query)
   at Lucene.Net.Search.BooleanQuery.Rewrite(IndexReader reader)
   at Lucene.Net.Search.BooleanQuery.Rewrite(IndexReader reader)
   at Lucene.Net.Search.BooleanQuery.Rewrite(IndexReader reader)
   at Lucene.Net.Search.IndexSearcher.Rewrite(Query original)
   at Lucene.Net.Search.Query.Weight(Searcher searcher)
   at Lucene.Net.Search.Searcher.Search(Query query, Filter filter, Collector results)
   at DotNetNuke.Services.Search.Internals.LuceneControllerImpl.Search(LuceneSearchContext searchContext)
   at DotNetNuke.Services.Search.Controllers.SearchControllerImpl.GetSecurityTrimmedResults(SearchQuery searchQuery, LuceneQuery luceneQuery)
   at DotNetNuke.Services.Search.Controllers.SearchControllerImpl.GetResults(SearchQuery searchQuery)
   at DotNetNuke.Services.Search.Controllers.SearchControllerImpl.SiteSearch(SearchQuery searchQuery)
   at DotNetNuke.Web.InternalServices.SearchServiceController.GetBasicViews(SearchQuery searchQuery, Int32& totalHits)
   at DotNetNuke.Web.InternalServices.SearchServiceController.GetGroupedBasicViews(SearchQuery query, SearchContentSource userSearchSource, Int32 portalId)
   at DotNetNuke.Web.InternalServices.SearchServiceController.Preview(String keywords, String culture, Int32 forceWild, Int32 portal)
```

the only workaround is restart that instance and search will back to work.

### PR DESCRIPTION
**Pull request by zyhfish**
fixes #2183

### CODE REVIEW

**Comment by sleupold** [comment]
what are the reasons for the search to throw an exception and does it really helps just to retry 5 times immediately in a row?

**Comment by zyhfish** [comment]
This is an edge case in azure instances, exception will thrown during migrate new instances, the drive need to prepare for a while, and in that time when request index files will get IOException, so need to close the reader and try again. actually its not retry immediately but wait for 100ms then retry. also the retry times can config to a larger number if we found the drive initialize will take longer time.

**Comment by ashishpd** [comment]
We are seeing this exception at customer's site. Please note most of the PR from @zyhfish come from real customer issues.

File: DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs [hunk 1]
```diff
@@ -274,41 +282,59 @@ public LuceneResults Search(LuceneSearchContext searchContext)
...
+                            TitleSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.TitleTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            BodySnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.BodyTag, searchContext.LuceneQuery.BodySnippetLength),
+                            DescriptionSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.DescriptionTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            TagSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.Tag, searchContext.LuceneQuery.TitleSnippetLength),
+                            AuthorSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.AuthorNameTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
+                        }).ToList();
                 return luceneResults;
+            }
+            catch (Exception ex)
```
Comment (tpluscode) [inline DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: If possible, please catch a more specific exception. You mention `IOException` in the ticket. Any other exception probably should not cause a retry, right?

File: DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs [hunk 2]
```diff
@@ -274,41 +282,59 @@ public LuceneResults Search(LuceneSearchContext searchContext)
...
-                        ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
-                    }).ToList();
-            return luceneResults;
+                DisposeReaders();
+                retryTimes++;
+                Logger.Error(ex);
+                Logger.Error($"Search Index Folder Is Not Available: {ex.Message}, Retry {retryTimes} time(s).");
+                Thread.Sleep(100);
+
+                return Search(searchContext, retryTimes);
```
Comment (tpluscode) [inline DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: Please do not call `Search` recursively from `try..catch`. Instead please wrap error handling in a loop and break out of that loop when the results have been successfully returned.

File: DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs (hunk 1 above)
Comment (zyhfish) [reply DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: yep, there have more exceptions but not only this one after deploy the patch and get the error log, have create several sub tasks to handle these errors.

Comment (tpluscode) [reply DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: Then please add a specific `catch` clause for each "expected exception". Let's start by catching the IOExcpetion in this PR and subsequent submissions will add more handling

File: DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs [hunk 3]
```diff
@@ -274,40 +277,61 @@ public LuceneResults Search(LuceneSearchContext searchContext)
...
+
+                    if (Logger.IsDebugEnabled)
+                    {
+                        var sb = GetSearcResultExplanation(searchContext.LuceneQuery, searchSecurityTrimmer.ScoreDocs, searcher);
+                        Logger.Trace(sb);
+                    }
 
-            //Page doesn't exist
-            if (luceneResults.TotalHits < minResults)
-                return luceneResults;
```
Comment (mitchelsellers) [inline DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: Why increase the nesting depth with this change from a return if the total hits were less than mins?

Comment (zyhfish) [reply DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: this is not a new change in this PR.

Comment (mitchelsellers) [reply DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: @zyhfish Yes, this is a change in this PR.  You removed the old process which returned early, to an inversed logic that increases the IL nesting depth of the code

Comment (tpluscode) [reply DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs]: It's amusing that with the logical expression reversed the comment above is not incorrect :)

**Pull request merged**